xds/src/test/java/io/grpc/xds/CdsLoadBalancerTest.java

```diff
@@ -325,6 +330,88 @@ public void handleCdsConfigs() throws Exception {
     assertThat(xdsClientRef.xdsClient).isNull();
   }
 
+  private static class FakeSocketAddress extends SocketAddress {
+    final String name;
+
+    FakeSocketAddress(String name) {
+      this.name = name;
+    }
+
+    @Override
+    public String toString() {
+      return "FakeSocketAddress-" + name;
+    }
+  }
+
+  @Test
+  public void handleCdsConfigs_withUpstreamTlsContext() throws Exception {
+    assertThat(xdsClient).isNull();
+
+    String lbConfigRaw1 = "{'cluster' : 'foo.googleapis.com'}".replace("'", "\"");
+    @SuppressWarnings("unchecked")
+    Map<String, ?> lbConfig1 = (Map<String, ?>) JsonParser.parse(lbConfigRaw1);
+    ResolvedAddresses resolvedAddresses1 =
+         ResolvedAddresses.newBuilder()
+             .setAddresses(ImmutableList.<EquivalentAddressGroup>of())
+             .setAttributes(
+                 Attributes.newBuilder()
+                     .set(ATTR_LOAD_BALANCING_CONFIG, lbConfig1)
+                     .set(XdsAttributes.XDS_CLIENT_REF, xdsClientRef)
+                     .build())
+             .build();
+    cdsLoadBalancer.handleResolvedAddresses(resolvedAddresses1);
+
+    ArgumentCaptor<ClusterWatcher> clusterWatcherCaptor1 = ArgumentCaptor.forClass(null);
+    verify(xdsClient).watchClusterData(eq("foo.googleapis.com"), clusterWatcherCaptor1.capture());
+
+    UpstreamTlsContext upstreamTlsContext =
+        UpstreamTlsContext.newBuilder()
+            .setCommonTlsContext(
+                CommonTlsContext.newBuilder()
+                    .addTlsCertificateSdsSecretConfigs(
+                        SdsSecretConfig.newBuilder().setName("cert-sds-name"))
+                    .setValidationContextSdsSecretConfig(
+                        SdsSecretConfig.newBuilder().setName("valid-sds-name")))
+            .build();
+
+    ClusterWatcher clusterWatcher1 = clusterWatcherCaptor1.getValue();
+    clusterWatcher1.onClusterChanged(
+        ClusterUpdate.newBuilder()
+            .setClusterName("foo.googleapis.com")
+            .setEdsServiceName("edsServiceFoo.googleapis.com")
+            .setLbPolicy("round_robin")
+            .setEnableLrs(false)
+            .setUpstreamTlsContext(upstreamTlsContext)
+            .build());
+
+    assertThat(edsLbHelpers).hasSize(1);
+    assertThat(edsLoadBalancers).hasSize(1);
+    Helper edsLbHelper1 = edsLbHelpers.poll();
+    assertThat(edsLbHelper1).isInstanceOf(EdsLoadBalancingHelper.class);
```

Comment:
We shouldn't assert this because this is an implementation detail that relies on the fact that `EdsLoadBalancingHelper` is the outermost forwarding helper, which may no longer be true if we were adding new features (with new forwarding helpers). As you are verifying the behavior of `createSubchannel` later, this line is not needed.

Reply:
But this is unit-testing which implies I am testing this unit as a white-box and the test is aware of the internal implementation detail and taking advantage of it to test various conditions. Once new features/implementation are added the test will also need to be updated.

Reply:
My point is this does not really test any behavior, just verifying the implementation, and if implementation changes, this will also need to be updated. So the line is not needed. While the behavior verification below does not need any change as long as the spec is unchanged.